### DISCUSSION TITLE
retrieve poa data with iotools get_pvgis_tmy

### DISCUSSION THREAD

**Discussion by ClemensReichel** [pvlib/pvlib-python#2036] [Q&A] [answered]
Is it possible to retrieve poa Data with get_pvgis_tmy or any other io tool?

**Comment by AdamRJensen** [comment]
The pvlib.iotools.get_pvgis_hourly returns poa data.

**Reply by AdamRJensen** [reply]
Otherwise, you can use any of the many transposition models (e.g. Perez or Hay Davies) to calculate possible from dni and dhi.

**Reply by ClemensReichel** [reply]
Thank you. But the TMY data doesn't come as poa, am I right? I didn't know about transposition models. How do they work?

**Reply by kandersolar** [reply, marked as answer]
The pvlib function `pvlib.irradiance.get_total_irradiance` is the primary means of estimating POA irradiance from DNI and DHI: https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.irradiance.get_total_irradiance.html

Here are some examples of calculating POA irradiance with pvlib: https://pvlib-python.readthedocs.io/en/stable/gallery/index.html#irradiance-transposition

For information about transposition models in general, this section of the PVPMC website may be helpful (check the sub-pages in the left sidebar as well): https://pvpmc.sandia.gov/modeling-guide/1-weather-design-inputs/plane-of-array-poa-irradiance/

**Reply by ClemensReichel** [reply]
Thank you! Is there a way to contact you? I am struggling quite hard with an apparently easy calculation, which has worked before:

    aoi = pvlib.irradiance.aoi(
        surface_tilt, surface_azimuth, solarpos.apparent_zenith, solarpos.azimuth)

    iam = pvlib.iam.ashrae(aoi)

    effective_irradiance = poa_data["poa_direct"] * iam + poa_data["poa_diffuse"]

    temp_cell = pvlib.temperature.faiman(poa_data["poa_global"], poa_data["temp_air"], poa_data["wind_speed"])
    
   The main problem is, that the effective irradiance only returns NaN and I despair of it

**Reply by kandersolar** [reply]
This forum is appropriate to work through such issues :)

To resolve your issue, I suggest moving backwards through your code and examining each variable sequentially to identify which calculation step is introducing the NaN.  For example, is the `aoi` NaN too?

**Reply by ClemensReichel** [reply]
Yes, I have checked that. But all the other variables contain valid data.
The only thing that I have noticed is that the return from the effective_irradiance has 17520 lines of NaN, while all other have the expected 8760 values for every hour per year.

**Reply by cwhanse** [reply]
aoi and poa_data likely have different indexes.

**Reply by AdamRJensen** [reply]
> aoi and poa_data likely have different indexes.

Potentially just different time zone information, which is a common culprit.

**Reply by ClemensReichel** [reply]
aoi and poa_data have the same but the iam doesn't. 
But how to change the index, since  the numpy.ndarray object of iam has no attribute "index"?

**Reply by ClemensReichel** [reply]
I am trying to work with tmy data. But since it gets the data for one year from different months/years, I am trying to process into it a randomly chosen year (2019)

     import pvlib
     import pandas as pd


    def process_pv_data(latitude, longitude, surface_tilt, surface_azimuth, start, end, location_name,
                    celltype, v_mp, i_mp, v_oc, i_sc, alpha_sc, beta_voc, gamma_pmp, cells_in_series, temp_ref,
                    inverter_model, modules_per_string, strings_per_inverter, eta_inv_nom, eta_inv_ref, pdc0,
                    number_of_inverters):
         pvgis_data, months_selected, meta, inputs = pvlib.iotools.get_pvgis_tmy(
             latitude=latitude,
             longitude=longitude,
             outputformat='json',
             usehorizon=True,
             userhorizon=None,
             startyear=2006,
             endyear=2019,
             map_variables=True,
             url='https://re.jrc.ec.europa.eu/api/v5_2/',
             timeout=30
         )
    pvgis_data.index = pd.date_range(start="2019-01-01 00:00", end="2019-12-31 23:00", freq="h", tz="Europe/Berlin")

    print(pvgis_data)
    solarpos = pvlib.location.Location(latitude=latitude, longitude=longitude,
                                       tz="Europe/Berlin", altitude=537, name=location_name).get_solarposition(
        times=pd.date_range(start=start, end=end, freq="h"))

    # Calculate total irradiance components
    total_irradiance = pvlib.irradiance.get_total_irradiance(
        surface_tilt=surface_tilt,
        surface_azimuth=surface_azimuth,
        solar_zenith=solarpos['zenith'],
        solar_azimuth=solarpos['azimuth'],
        dni=pvgis_data['dni'],
        ghi=pvgis_data['ghi'],
        dhi=pvgis_data['dhi'],
        albedo=0.25,  # Ground albedo value
        model='isotropic'  # You can choose a different model if needed
    )

    total_irradiance['wind_speed'] = pvgis_data['wind_speed']
    total_irradiance['temp_air'] = pvgis_data['temp_air']

    # Extract only the desired columns
    selected_columns = ['poa_global', 'poa_direct', 'poa_diffuse', 'poa_sky_diffuse', 'poa_ground_diffuse',
                        "wind_speed", "temp_air"]
    poa_data = total_irradiance[selected_columns]


    poa_data.to_csv(f"poa_data_{location_name}.csv")

    poa_data = pd.read_csv(f"poa_data_{location_name}.csv", index_col=0)

    aoi = pvlib.irradiance.aoi(
        surface_tilt, surface_azimuth, solarpos.apparent_zenith, solarpos.azimuth)

    iam = pvlib.iam.ashrae(aoi)


    effective_irradiance = poa_data["poa_direct"] * iam + poa_data["poa_diffuse"]


    temp_cell = pvlib.temperature.faiman(poa_data["poa_global"], poa_data["temp_air"], poa_data["wind_speed"])


The locations and the process of the pv data is defined later
